src/main/java/frc/trigon/robot/subsystems/swerve/swervemodule/SwerveModule.java

```diff
@@ -11,144 +11,148 @@
-        this.wheelDiameterMeters = wheelDiameterMeters;
+        steerEncoder = new CANcoderEncoder(moduleID + 4, "Module" + moduleID + "Steer", RobotConstants.CANIVORE_NAME);
+
         configureHardware(offsetRotations);
     }
 
-    void setTargetDriveMotorCurrent(double targetCurrent) {
-        driveMotor.setControl(driveTorqueCurrentFOCRequest.withOutput(targetCurrent));
-    }
+    /**
+     * Sets the target angle and speed of the module from a target module state for the angle, and a target current for the drive speed.
+     * This is used for PathPlanner to follow a path using feedforwards as well as chassisSpeeds.
+     *
+     * @param swerveModuleState the target module state. Only uses the angles
+     * @param targetCurrent     the target current of the drive motor
+     */
+    public void setTargetState(SwerveModuleState swerveModuleState, double targetCurrent) {
+        final Rotation2d unoptimizedAngle = swerveModuleState.angle;
+        swerveModuleState.optimize(getCurrentAngle());
+        setTargetAngle(swerveModuleState.angle);
 
-    void udpateDriveMotorLog(SysIdRoutineLog log) {
-        log.motor("Module" + driveMotor.getID() + "Drive")
-                .angularPosition(Units.Rotations.of(driveMotor.getSignal(TalonFXSignal.POSITION)))
-                .angularVelocity(Units.RotationsPerSecond.of(driveMotor.getSignal(TalonFXSignal.VELOCITY)))
-                .voltage(Units.Volts.of(driveMotor.getSignal(TalonFXSignal.TORQUE_CURRENT)));
+        if (!swerveModuleState.angle.equals(unoptimizedAngle)) {
+            setDriveMotorTargetCurrent(-targetCurrent);
+            return;
+        }
+        setDriveMotorTargetCurrent(targetCurrent);
     }
 
-    void stop() {
-        driveMotor.stopMotor();
-        steerMotor.stopMotor();
+    public void setTargetState(SwerveModuleState targetState) {
+        targetState.optimize(getCurrentAngle());
+        this.targetState = targetState;
+        setTargetAngle(targetState.angle);
+        setTargetVelocity(targetState.speedMetersPerSecond, targetState.angle);
     }
 
-    void setBrake(boolean brake) {
+    public void setBrake(boolean brake) {
         driveMotor.setBrake(brake);
         steerMotor.setBrake(brake);
     }
 
-    void update() {
+    public void updateLog(SysIdRoutineLog log) {
+        log.motor("Module" + driveMotor.getID() + "Drive")
+                .angularPosition(Units.Rotations.of(driveMotor.getSignal(TalonFXSignal.POSITION)))
+                .angularVelocity(Units.RotationsPerSecond.of(driveMotor.getSignal(TalonFXSignal.VELOCITY)))
+                .voltage(Units.Volts.of(driveMotor.getSignal(TalonFXSignal.TORQUE_CURRENT)));
+    }
+
+    /**
+     * Updates the swerve module. Should be called periodically.
+     * This method updates the hardware, and updates the motor positions variable.
+     * We save the positions to a variable instead of getting them directly because the signals update at a higher frequency than the main code loop.
```

Comment:
```suggestion
     * We save their positions to variables instead of getting them directly because their signals update at a higher frequency than the main code loop.
```